### PR TITLE
[Coop] InterlockedCompare/Exchange (object/T)

### PR DESCRIPTION
**Pull request by jaykrell**
- Usually intrinsified.
 - Always?
 - Avoid coop handles by manually always using ref to object.
 - For efficiency on input, and efficiency wrt MonoError.
 - For correctness on output (due to aliasing in the wrappers).
 - To probably avoid assertion failure in runtime around icall marshaling of generics.
 - Unclear on location.

### CODE REVIEW

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@akoeplinger look at tarball build?

File: netcore/System.Private.CoreLib/src/System.Threading/Interlocked.cs [hunk 1]
```diff
@@ -13,7 +13,26 @@ public static class Interlocked
...
+			// location1's treatment is unclear. But note that passing it by handle would be incorrect,
+			// as it would use a local alias, which the coop marshaling does, to avoid the unclarity here,
+			// that of a ref being to a managed frame vs. a native frame. Perhaps that could be revisited.
+			//
+			// So there a hole here, that of calling this function with location1 being in a native frame.
+			// Usually it will be to a field, static or not, and not even to managed stack.
+			//
+			// This is usually intrinsified. Ideally it is always intrinisified.
+			//
+			object result = null;
```
Comment (jaykrell) [inline netcore/System.Private.CoreLib/src/System.Threading/Interlocked.cs]: Ideally the parameters and locals could be marked volatile. In that sense this is not equivalent to the icall wrapper. Maybe should fix that. And try “fixed” and address-of on the generic?

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by kg** [comment]
Can you provide some clarity on what this PR does? I'm guessing that moving more of this logic into managed code reduces the amount of time spent in runtime/natives and allows more of the old runtime stuff to happen in intrinsics. So is it an optimization or does it also increase correctness?

Does it need new tests? I'm not sure how good our coverage is for interlocked/atomics.

**Comment by jaykrell** [comment]
None of the above. It is coop conversion.

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

File: mcs/class/corlib/System.Threading/Interlocked.cs [hunk 2]
```diff
@@ -92,9 +122,28 @@ public static partial class Interlocked
...
 		[MethodImplAttribute(MethodImplOptions.InternalCall)]
+		[return: NotNullIfNotNull("location1")]
+		extern static void CompareExchange_T (ref T location1, ref T value, ref T comparand, ref T result) where T : class?;
+
+		[ComVisible (false)]
 		[ReliabilityContract (Consistency.WillNotCorruptState, Cer.Success)]
-		public extern static T CompareExchange<T> (ref T location1, T value, T comparand) where T:class;
+		[return: NotNullIfNotNull("location1")]
+		public static T CompareExchange<T> (ref T location1, T value, T comparand) where T : class?
+		{
```
Comment (jaykrell) [inline mcs/class/corlib/System.Threading/Interlocked.cs]: Without the ability to make locals volatile, this seems fragile under heavy optimization.

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
Git clean failing again.

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
Tar ball finally

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by kg** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build  Windows x64 FullAOT

**Comment by jaykrell** [comment]
@monojenkins build Windows x64 FullAOT

**Comment by jaykrell** [comment]
@monojenkins build Windows x64 FullAOT+LLVM

**Comment by jaykrell** [comment]
@monojenkins build  Linux x64 FullAOT+Interpreter

**Comment by jaykrell** [comment]
@monojenkins build  Linux x64 FullAOT+LLVM

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
git clean failing again

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
netcore is very finicky here due to nullable stuff

**Comment by jaykrell** [comment]
Will try to repro all this locally.

**Comment by jaykrell** [comment]
Problem was obvious with quick local repro and printf debugging.
And not surpring.
The signature checkin aot-runtime.c, I had looked at it, and decided it'd work, but I was wrong.
The type changed from underlying type being a reference type, to actual type being a reference type, and underlying type no longer being so. So fix is to check actual and not underlying type.
Strengthened it while there -- or made it more fragile, depending on perspective.

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
Better but still some failures. Let's try no signature check in aot-runtime.c and rename the function to be sure (not sure it is already down generic path here).

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
And put the _T in aot-compiler.c too.

**Comment by jaykrell** [comment]
Rereview post-approval changes or ok?

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
I thought only failures were bogus api diff.

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
(Linux armv7 interpreter)

**Comment by vargaz** [comment]
@monojenkins commit apidiff

**Comment by jaykrell** [comment]
Apidiff resets CI?

**Comment by jaykrell** [comment]
Squash?

**Comment by lewing** [comment]
This appears to have broken profiled aot for wasm

**Comment by jaykrell** [comment]
Something that CI does not test?

**Pull request merged**